rasterio/_warp.pyx

```diff
@@ -322,8 +322,7 @@ def _reproject(
 
             finally:
                 CPLFree(srcwkt)
-                OSRRelease(osr)
-                osr = NULL
+                _safe_osr_release(osr)
```

Comment:
@jdmcbr I found the crasher: we need to restore `osr = NULL` after line 325 (and after 416, below) because we're reusing the osr variable. Reusing the variable is clearly the root of the problem, we should probably define one for the src crs and one for the dst crs in `_reproject()` to be more future-proof.